### PR TITLE
Launchpad Revival

### PR DESCRIPTION
**Pull request by XDTM**
## About The Pull Request

QoL improvements for launchpads, heavy buffs as well as the introduction of a countermeasure.
In detail:
- **Changed the UI for the launchpad console into TGUI.**
- Added manual set X and Y buttons as an alternative to arrows.
- **Diagnostic HUDs can now see the current target turf of active launchpads. They can also see if they are currently sending or receiving.**
- **Launchpads now get an additional 15 range with every manipulator upgrade, instead of 1.** Now it can range from 15 tiles up to 60.

Note: before you ask me to atomize the buff, i don't think that it's a good idea without the counterplay of being able to see the indicator, and it's pretty unusable without the manual Set X/Y buttons; it all has to go together in this case.

Screenshots:
New interface and targeting reticles:

Launchpad Controls:

Sending:

Receiving:

## Why It's Good For The Game

Makes launchpads more usable, while also giving people a way to counteract it. Launchpads will always have a 3.5 second teleport delay no matter the upgrades, so if you spot the indicator hovering over an armory locker, for example, you could just take its place and arrest the intruding scientists.
On the other hand, diag HUDs are fairly rare outside of robotics and silicons, so a launchpad user can go unnoticed if they're being careful.

## Changelog
:cl: XDTM
add: Launchpads now gain 15 range instead of 1 when upgraded, up to a maximum of 60.
add: Launchpads now have a targeting reticle which is visible on Diagnostic HUD. It also indicates if the launchpad is currently sending or receiving.
tweak: Launchpad Consoles now use TGUI and are far more responsive. They can also set their coordinate offsets manually instead of only using arrows.
/:cl:

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
Thanks for preemptively clarifying it's relative rather than the turf's X/Y coord in the input's, that was my first question when I saw the X/Y.

**Comment by Akrilla** [comment]
Solid change

**Comment by Kollock5** [comment]
maximum of 60? you can raid the armory from the vacant office with a range of 60... you might wanna overthink that...

**Comment by ExcessiveUseOfCobblestone** [comment]
at T4...

**Comment by Tlaltecuhtli** [comment]
>Changed the UI for the launchpad console into TGUI.

**Comment by Nabski89** [comment]
Does the UI change touch the briefcase launchpad too?

**Comment by XDTM** [comment]
The briefcase already had a TGUI interface, so i didn't touch it. It is missing the Set X/Y buttons that i added to the console, but since it has a fairly limited range it shouldn't matter. 
More importantly, it does have a marker visible to DiagHUDs. I'm considering potentially making an exception to allow for more stealth, though.

**Comment by Swiftkor** [comment]
oh boy now I can teleport my rev bombs directly into brig

**Comment by Shadowflame909** [comment]
More shit for science to do, yippie.

I just hope it's not useless at round-start like nanites is. That's why it's pretty much never used.

**Comment by py01** [comment]
Does this have any effect on the traitor's breficase launchpad?

**Comment by GunHog** [comment]
One more visibility thing you could consider is having it draw a teleport beam from the target to the source, the same way the Bluespace Rapid Part Exchange Device and Bag of Holding do, during the 3.5 second charging timer. Most of the time the source will be Science, but this will let anyone trace the signal's direction. 

If that is too much, another suggestion is to indicate direction or path somehow on the HUD. The reasoning is that while players can see the reticule, they might not be able to figure out the general direction of the action.

Even so, it is a brilliant addition as is :+1:

**Comment by XDTM** [comment]
I feel like pointing to the teleporter might be a bit too unsafe for the user if they're using it for crime. If it turns out to be too strong it can always be nerfed again, but i want to bring it in the spotlight first.

**Comment by ATH1909** [comment]
I do like the "indication of what direction the launchpad is, but only on the/visible with a diagnostic HUD" idea, though. If this ends up needing a nerf, we should go with that one.

**Comment by davethwave** [comment]
Welp, just needs to fix fix the conflict then hope someone merges this gem.

File: code/game/machinery/computer/launchpad_control.dm [hunk 1]
```diff
@@ -38,113 +38,180 @@
...
\ No newline at end of file
+	switch(action)
+		if("return")
+			screen = "select"
+			. = TRUE
+		if("select_pad")
+			current_pad = get_pad(text2num(params["id"]))
+			screen = "control"
+			. = TRUE
+		if("right")
```
Comment (AnturK) [inline code/game/machinery/computer/launchpad_control.dm]: The redundancy of these cases is causing me physical pain. Add dx/dy params to the act then you can just do it in one go please.

**Review by AnturK — COMMENTED** [review]
Looks alright, needs tgui update. And one code improvement suggestion.

**Pull request merged**